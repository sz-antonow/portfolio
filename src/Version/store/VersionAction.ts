import { Dispatch } from 'redux';

import {VersionConst} from "./VersionConst";
import {VersionApi} from "../api/VersionApi";

// import { VersionApi } from 'src/module/Version/api/VersionApi';
// import { VersionConst } from 'src/module/Version/store/VersionConst';
// import { AsyncActionCreator } from 'src/store/async';

export namespace VersionAction {
  export const fetchVersionActions: AsyncActionCreator<any> = AsyncActionCreator.init(VersionConst.FETCH_VERSION);

  export const fetchVersion = () => (dispatch: Dispatch<any>) => {
    dispatch(fetchVersionActions.pending());

    return VersionApi.getVersion()
      .then((data) => {
        dispatch(fetchVersionActions.resolved(data));
      })
      .catch((error) => dispatch(fetchVersionActions.rejected(error)));
  };
}
