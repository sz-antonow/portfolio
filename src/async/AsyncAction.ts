import { ActionMeta } from 'redux-actions';



export type AsyncAction<Payload> = ActionMeta<Payload, AsyncActionMeta>;

export namespace AsyncAction {
  /**
   * Creates action type
   * from given action type and current async status
   * @param type
   * @param asyncStatus
   */
  export const createType = (type: string, asyncStatus: AsyncStatus): string => `${type}_${asyncStatus}`;

  export type CreateTypes = {
    VOID: string;
    PENDING: string;
    RESOLVED: string;
    REJECTED: string;
  };
  export const createTypes = (type: string): CreateTypes => ({
    VOID: createType(type, AsyncStatus.VOID),
    PENDING: createType(type, AsyncStatus.PENDING),
    RESOLVED: createType(type, AsyncStatus.RESOLVED),
    REJECTED: createType(type, AsyncStatus.REJECTED)
  });

  /**
   * Action init for async actions init
   * returns action marked as async
   * @param type
   * @param asyncStatus
   * @param payload
   */
  export function create<Payload>(type: string, asyncStatus: AsyncStatus, payload: Payload): AsyncAction<Payload> {
    const action = {
      type: createType(type, asyncStatus),
      payload,
      meta: {
        type,
        asyncStatus: asyncStatus
      }
    };

    return asyncStatus === AsyncStatus.REJECTED ? { ...action, error: true } : action;
  }
}
