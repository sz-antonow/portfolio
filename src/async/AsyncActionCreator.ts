import { AsyncStatus } from 'common/src/enum';
import { AsyncAction } from 'common/src/store/async/AsyncAction';

export type AsyncActionCreator<Payload = any> = Readonly<{
  pending: (payload?: Payload) => AsyncAction<Payload>;
  resolved: (payload?: Payload) => AsyncAction<Payload>;
  rejected: (error: Payload) => AsyncAction<Payload>;
}>;

export namespace AsyncActionCreator {
  /**
   * Creates action creators
   * for all async statuses
   * @param type
   */
  export function init<Payload>(type: string): AsyncActionCreator<Payload> {
    return {
      pending: (payload: Payload = null) => AsyncAction.create<Payload>(type, AsyncStatus.PENDING, payload),
      resolved: (payload: Payload) => AsyncAction.create<Payload>(type, AsyncStatus.RESOLVED, payload),
      rejected: (error: Payload) => AsyncAction.create<Payload>(type, AsyncStatus.REJECTED, error)
    };
  }
}
