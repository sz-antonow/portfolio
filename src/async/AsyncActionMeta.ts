import { AsyncStatus } from 'app/common/src/enum';

export type AsyncActionMeta = Readonly<{
  type: string;
  asyncStatus: AsyncStatus;
}>;
