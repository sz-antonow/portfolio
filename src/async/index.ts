export { AsyncAction } from './AsyncAction';
import { AsyncActionMeta as AsAcM } from './AsyncActionMeta';
export { AsyncActionCreator } from './AsyncActionCreator';
export { AsyncState } from './AsyncState';
export { AsyncSelector } from './AsyncSelector';

export type AsyncActionMeta = AsAcM;
