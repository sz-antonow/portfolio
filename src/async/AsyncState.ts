import { AxiosError } from 'axios';

import { AsyncStatus } from 'app/common/src/enum';

export interface AsyncState<Data> {
  status: AsyncStatus;
  data: Data;
  error?: AxiosError;
}

export namespace AsyncState {
  export type Create<Data, State extends AsyncState<Data> = AsyncState<Data>> = (
    data: Data,
    error?: AxiosError
  ) => State;

  export function create<Data>(status: AsyncStatus): Create<Data> {
    return (data: Data, error?: AxiosError): AsyncState<Data> => (error ? { data, status, error } : { status, data });
  }

  export function pending<Data>(data: Data = null): AsyncState<any> {
    return create(AsyncStatus.PENDING)(data);
  }

  export function resolved<Data>(data: Data): AsyncState<any> {
    return create(AsyncStatus.RESOLVED)(data);
  }

  export function rejected<Data>(data?: Data, error?: any): AsyncState<any> {
    return create(AsyncStatus.REJECTED)(data, error);
  }

  export function initial<Data>(data: Data = null): AsyncState<any> {
    return create(AsyncStatus.VOID)(data);
  }
}
