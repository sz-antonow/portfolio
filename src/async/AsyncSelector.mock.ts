import { AsyncStatus } from 'common/src/enum';
import { AsyncState } from 'common/src/store/async/AsyncState';

export const axiosError = {
  config: {},
  name: '',
  message: '',
  response: {
    status: 409,
    statusText: 'Conflict',
    data: {
      message: 'Custom backend error message'
    },
    headers: {},
    config: {}
  },
  isAxiosError: true,
  toJSON: () => {
    return {};
  }
};

export const asyncStateResolved: AsyncState<string[]> = {
  status: AsyncStatus.RESOLVED,
  data: ['foo', 'bar']
};

export const asyncStatePending: AsyncState<string[]> = {
  status: AsyncStatus.PENDING,
  data: []
};

export const asyncStateRejected: AsyncState<string[]> = {
  status: AsyncStatus.REJECTED,
  data: [],
  error: axiosError
};
