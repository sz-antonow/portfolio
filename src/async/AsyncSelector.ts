import { AxiosError } from 'axios';
import { createSelector, Selector } from 'reselect';

import { AsyncStatus } from 'app/common/src/enum';
import { AsyncState } from 'app/common/src/store/async/AsyncState';

export namespace AsyncSelector {
  export function selectStatus<Data>(asyncState: AsyncState<Data>): AsyncStatus {
    return asyncState.status;
  }

  export function selectData<Data>(asyncState: AsyncState<Data>): Data {
    return asyncState.data;
  }

  export function selectError<Data>(asyncState: AsyncState<Data>): AxiosError {
    return asyncState.error;
  }

  export function selectErrorResponse<Data>(asyncState: AsyncState<Data>) {
    if (selectError(asyncState)) {
      return asyncState.error.response;
    }
  }

  export function selectErrorMessage<Data>(asyncState: AsyncState<Data>): string {
    if (selectErrorResponse(asyncState)) {
      return asyncState.error.response.data.message;
    }
  }

  export function selectErrorCode<Data>(asyncState: AsyncState<Data>): number | string {
    if (selectErrorResponse(asyncState)) {
      return asyncState.error.response.data.code;
    }
  }

  export function selectDomainStatus<State, Data>(domainSelector: Selector<State, AsyncState<Data>>) {
    return createSelector(
      domainSelector,
      (domain: AsyncState<Data>): AsyncStatus => selectStatus(domain)
    );
  }

  export function selectDomainData<State, Data>(domainSelector: Selector<State, AsyncState<Data>>) {
    return createSelector(
      domainSelector,
      (domain: AsyncState<Data>): Data => selectData(domain)
    );
  }
}
