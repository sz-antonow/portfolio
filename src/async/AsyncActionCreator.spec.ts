import { AsyncStatus } from 'common/src/enum';
import { AsyncAction } from 'common/src/store/async';
import { AsyncActionCreator } from 'common/src/store/async';

describe('AsyncActionCreator', () => {
  it('should return PENDING action from async actions init', () => {
    const type = 'ACTION_TYPE';
    const asyncActionsCreator = AsyncActionCreator.init(type);
    const pendingAction = AsyncAction.create(type, AsyncStatus.PENDING, null);

    expect(asyncActionsCreator.pending()).toEqual(pendingAction);
  });

  it('should return RESOLVED action from async actions init', () => {
    const type = 'ACTION_TYPE';
    const asyncActionsCreator = AsyncActionCreator.init(type);
    const payload = ['foo'];
    const resolvedAction = AsyncAction.create(type, AsyncStatus.RESOLVED, payload);

    expect(asyncActionsCreator.resolved(payload)).toEqual(resolvedAction);
  });

  it('should return rejected action from async actions init', () => {
    const type = 'ACTION_TYPE';
    const asyncActionsCreator = AsyncActionCreator.init(type);
    const rejectedAction = AsyncAction.create(type, AsyncStatus.REJECTED, new Error());

    expect(asyncActionsCreator.rejected(new Error())).toEqual(rejectedAction);
  });

  it('should create async actions creators', () => {
    const type = 'ACTION_TYPE';
    const asyncActionsCreator = AsyncActionCreator.init(type);

    expect(asyncActionsCreator).toMatchSnapshot();
  });
});
