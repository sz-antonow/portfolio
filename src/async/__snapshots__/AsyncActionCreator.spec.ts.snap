// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncActionCreator should create async actions creators 1`] = `
Object {
  "pending": [Function],
  "rejected": [Function],
  "resolved": [Function],
}
`;
